python/mlcroissant/mlcroissant/_src/operation_graph/operations/parse_json.py

```diff
@@ -1,20 +1,223 @@
     """Parsed all JSONs defined in the fields of RecordSet and outputs a pd.DF."""
     series = {}
     for field in fields:
         json_path = field.source.extract.json_path
-        if json_path is None:
+        if not json_path:
             continue
-        jsonpath_expression = jsonpath_rw.parse(json_path)
-        values = [match.value for match in jsonpath_expression.find(json)]
-        series[json_path] = values
+        expr = jsonpath_rw.parse(json_path)
+        vals = []
+        for match in expr.find(json_obj):
+            v = match.value
+            # If we got back a one‐item list, unwrap it.
+            if isinstance(v, list) and len(v) == 1:
+                v = v[0]
+            vals.append(v)
+        series[json_path] = vals
     return pd.DataFrame(series)
+
+
+class JsonReader:
+    """Parser for JSON files, supporting both JSONPath and JMESPath expressions."""
+
+    def __init__(self, fields: tuple[Field, ...]):
+        """Initializes the parser with a tuple of Field objects.
+
+        Args:
+            fields (tuple[Field, ...]): A tuple of Field objects, each containing
+            a source with a JSON path to extract.
+
+        The constructor builds a list of tuples for each field with a valid
+        JSON path:
+            - The original JSON path string.
+            - The engine used for evaluation ("jsonpath" for recursive-descent
+              paths, "jmespath" for simple direct paths).
+            - The compiled expression object for efficient evaluation.
+
+        Fields without a JSON path are skipped.
+        """
+        import jmespath
+
+        # Build a list of (original_jsonpath, engine, compiled_expr).
+        self.exprs: list[tuple[str, str, Any]] = []
+        for field in fields:
+            json_path = field.source.extract.json_path
+            if not json_path:
+                continue
+
+            # Decide whether this path can be JMESPath or needs full JSONPath.
+            stripped = json_path.lstrip("$.")
+            if ".." in json_path:
+                # Uses recursive‐descent → fall back to jsonpath_ng.
+                expr = jsonpath_rw.parse(json_path)
+                engine = "jsonpath"
+            else:
+                # Simple direct path → use JMESPath.
+                expr = jmespath.compile(stripped)
+                engine = "jmespath"
+
+            self.exprs.append((json_path, engine, expr))
+        self.fields = fields
+
+    def parse(self, fh: TextIO) -> pd.DataFrame:
+        """Parses a JSON file-like object and extracts data into a pandas DataFrame.
+
+        Args:
+            fh (TextIO): A file-like object containing JSON data.
+
+        Returns:
+            pd.DataFrame: DataFrame with extracted data,
+            where each column corresponds to an expression.
+        """
+        import orjson
+
+        # Load entire JSON file (could be a list or a single dict).
+        raw = fh.read()
+        data = orjson.loads(raw)
```

Comment:
You can see here an example of how to lazily load a library: https://github.com/mlcommons/croissant/pull/835/commits/4fbd358f14ae48bd217ad5beb69e6683c0b9e384